hmt_escrow/storage.py

```diff
@@ -28,10 +47,39 @@ def _connect(host: str, port: int) -> Client:
         raise e
 
 
-IPFS_CLIENT = _connect(IPFS_HOST, IPFS_PORT)
+def _connect_s3():
+    try:
+        return boto3.client("s3",
+                            aws_access_key_id=ESCROW_AWS_ACCESS_KEY_ID,
+                            aws_secret_access_key=ESCROW_AWS_SECRET_ACCESS_KEY,
+                            endpoint_url=ESCROW_ENDPOINT_URL,
+                            region_name=ESCROW_AWS_REGION)
+    except Exception as e:
+        LOG.error(f"Connection with S3 failed because of: {e}")
+        raise e
+
+
+def _put_object_kwargs(s3_hash, encrypted_msg):
+    """Make sure we can do local development with minio without using KMS."""
 
+    # If LOCAL is set to True, we don't use KMS.
+    if LOCAL:
+        return {
+            "Bucket": ESCROW_BUCKETNAME,
+            "Key": s3_hash,
+            "Body": encrypted_msg
+        }
+    else:
+        return {
+            "ServerSideEncryption": "aws:kms",
+            "SSEKMSKeyId": ESCROW_AWS_KMS,
+            "Bucket": ESCROW_BUCKETNAME,
+            "Key": s3_hash,
+            "Body": encrypted_msg
+        }
 
-def download(key: str, private_key: bytes) -> Dict:
+
+def download(key: str, private_key: bytes, s3: bool = True) -> Dict:
```

Comment:
I'm curious why we need the s3 arg here.. can't we support both by just looking at the key itself?  (ie if `s3://...` then use s3, if not, default to ipfs?). Supports both and allows for backward compatibility while we roll this out, and doesn't change the API

Reply:
The key here would be the hash we store in the contract and inside the bucket with the following encryption:
`hash_ = hashlib.sha1(manifest_.encode('utf-8')).hexdigest()` which will act as an index inside the bucket. 

The hash looks something like this: `3399620f41cd85c362e4b59a18c15e23cf746c66`

Currently this doesn't change the API since we would be defaulting to s3, it will only change if we ever want to switch back to IPFS.

Reply:
It does change the api though, since you're defaulting to something that is different from what it was before.

Reply:
I guess an `s3` prefix could be added but that then again will be exposed to the public via the contract, anyone could see it. 

I see the point though of having a prefix that could be used to automatically choose between ipfs and s3 when downloading. When uploading we would however have to define the strategy (s3 / ipfs).

Reply:
The outside API doesn't change, just the end destination of the content (ipfs -> s3). We would not need to change anything in the code.

Reply:
The interface doesn't change, but the behavior does.  its a breaking change since I can't just upgrade this in place in my existing code, I need to go specify `s3=False` for the same behavior

> When uploading we would however have to define the strategy (s3 / ipfs).

That's true, I would expect the default to be ipfs and if s3=true then do the s3 upload.

Could always just increase the minor version as well; my bigger concern is just making it backward compatible so we dont have cases where as soon as we upgrade we can't do anything with the old jobs that are still ipfs based

Reply:
What if we added two functions: `s3_upload` `s3_download`? Would be probably the clearest without more advanced if logic while guaranteeing that we maintain backward compatibility?

Reply:
The other option would be to default to IPFS as you mentioned.

Reply:
With the two functions you have a similar problem .. you have to have the calling code differentiate between "is it s3 or is it ipfs" which would work just need to do more on the calling side.

Reply:
Changes made:
1. `upload` constructs an `s3` prefix to the `manifest_url` if s3 has been chosen. Defaults to ipfs
2. `download` automatically determines whether to choose ipfs or s3 based on the prefix